src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.CopyFile.OSX.cs

```diff
@@ -0,0 +1,148 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.IO
+{
+    internal static partial class FileSystem
+    {
+        public static partial void CopyFile(string sourceFullPath, string destFullPath, bool overwrite)
+        {
+            // Attempt to clone the file:
+
+            // Helper function to throw an error for copying onto self
+            [StackTraceHidden]
+            void CopyOntoSelfError()
+            {
+                // Throw an appropriate error
+                if (overwrite) throw new IOException(SR.Format(SR.IO_SharingViolation_File, destFullPath));
+                else throw new IOException(SR.Format(SR.IO_FileExists_Name, destFullPath));
+            }
+
+            // Helper function to throw an error when the destination exists and overwrite = false.
+            [StackTraceHidden]
+            static void DestinationExistsError()
+            {
+                throw Interop.GetExceptionForIoErrno(new Interop.ErrorInfo(Interop.Error.EEXIST));
+            }
+
+            // Fail fast for blatantly copying onto self
+            if (sourceFullPath == destFullPath)
+            {
+                CopyOntoSelfError();
+            }
+
+            // Start the file copy and prepare for finalization
+            StartedCopyFileState startedCopyFile = StartCopyFile(sourceFullPath, destFullPath, overwrite, openDst: false);
```

Comment:
Any particular reason not to use `using` here?

Reply:
@filipnavara this line modifies the instance, which is disallowed with `using`.
https://github.com/dotnet/runtime/blob/776f500f41b83424e83b081982fd087e32f46131/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.CopyFile.OSX.cs#L136